polkadot/xcm/procedural/src/builder_pattern.rs

```diff
@@ -17,51 +17,53 @@
 //! Derive macro for creating XCMs with a builder pattern
 
 use inflector::Inflector;
-use proc_macro::TokenStream;
 use proc_macro2::TokenStream as TokenStream2;
 use quote::{format_ident, quote};
 use syn::{
-	parse_macro_input, Data, DeriveInput, Error, Expr, ExprLit, Fields, Lit, Meta, MetaNameValue,
+	Data, DeriveInput, Error, Expr, ExprLit, Fields, Lit, Meta, MetaNameValue,
+	Result, Variant, Ident, DataEnum, parse_quote,
 };
 
-pub fn derive(input: TokenStream) -> TokenStream {
-	let input = parse_macro_input!(input as DeriveInput);
-	let builder_impl = match &input.data {
-		Data::Enum(data_enum) => generate_methods_for_enum(input.ident, data_enum),
-		_ =>
-			return Error::new_spanned(&input, "Expected the `Instruction` enum")
-				.to_compile_error()
-				.into(),
+pub fn derive(input: DeriveInput) -> Result<TokenStream2> {
+	let data_enum = match &input.data {
+		Data::Enum(data_enum) => data_enum,
+		_ => return Err(Error::new_spanned(&input, "Expected the `Instruction` enum"))
 	};
+	let builder_raw_impl = generate_builder_raw_impl(&input.ident, data_enum);
+	let builder_impl = generate_builder_impl(&input.ident, data_enum)?;
+	let builder_unpaid_impl = generate_builder_unpaid_impl(&input.ident, data_enum)?;
 	let output = quote! {
-		pub struct XcmBuilder<Call>(Vec<Instruction<Call>>);
+		/// Type used to build XCM programs that require fee payment
+		pub struct XcmBuilder<Call>(pub(crate) Vec<Instruction<Call>>);
+		/// Type used to build XCM programs that require explicitly stating no fees need to be paid
+		pub struct XcmBuilderUnpaid<Call>(pub(crate) Vec<Instruction<Call>>);
+		/// Type used to build arbitrary XCM, without restrictions.
+		/// Should only be used when you know what you're doing, for experimenting or for learning/teaching purposes.
+		pub struct XcmBuilderRaw<Call>(pub(crate) Vec<Instruction<Call>>);
```

Comment:
These types should probably be generated conditionally (under a feature flag maybe ?) as they are not supposed to be used in production.

Instead of creating multiple types, I would create the same one with different functionality based on an attribute in the macro.

For example, `dev_mode` in pallet construction allows you to omit certain things, that would complain as soon as you remove it for production. Can we do something similar here?

Reply:
I was thinking they were supposed to be used in production. Sure, we could just use the raw one but the paid and unpaid are really good for making sure we don't forget to pay or explicitly show we are not paying. I don't see a reason not to allow these for everyone using the xcm crate

Reply:
But there's a comment there stating `Should only be used when you know what you're doing, for experimenting or for learning/teaching purposes.` ?

Reply:
Oh yeah, we could probably feature flag only the raw one, you're right. That'd be good. However, my other builders piggyback on that one, so probably do it private and feature flag a public version?

Reply:
You can also make use of typestates, which would obviate the need to have different structs that look almost the same (albeit you still need some unit structs).